File: src/store.py
import json
from annoy import AnnoyIndex
import os
from typing import List, Dict, Optional


class VectorStore:
    def __init__(self, dim: int, index_path: str):
        self.dim = dim
        self.index_path = index_path
        self.meta_path = index_path + ".meta.json"
        self.index = AnnoyIndex(dim, "angular")
        self._id_to_meta: Dict[int, Dict] = {}
        self._next_id = 0

    def add_documents(self, docs: List[Dict[str, str]], embeddings):
        # docs: list of {'text':..., 'id': optional}
        for doc, emb in zip(docs, embeddings):
            idx = self._next_id
            self.index.add_item(idx, emb.tolist() if hasattr(emb, "tolist") else emb)
            self._id_to_meta[idx] = {"text": doc.get("text"), "source_id": doc.get("id")}
            self._next_id += 1

    def build(self, n_trees: int = 10):
        self.index.build(n_trees)

    def save(self):
        os.makedirs(os.path.dirname(self.index_path) or ".", exist_ok=True)
        self.index.save(self.index_path)
        with open(self.meta_path, "w", encoding="utf-8") as f:
            json.dump({"meta": self._id_to_meta, "next_id": self._next_id}, f)

    @classmethod
    def load(cls, dim: int, index_path: str) -> "VectorStore":
        vs = cls(dim, index_path)
        if os.path.exists(index_path):
            vs.index.load(index_path)
        if os.path.exists(vs.meta_path):
            with open(vs.meta_path, "r", encoding="utf-8") as f:
                data = json.load(f)
                vs._id_to_meta = {int(k): v for k, v in data.get("meta", {}).items()}
                vs._next_id = int(data.get("next_id", 0))
        return vs

    def query(self, embedding, top_k: int = 5):
        ids, distances = self.index.get_nns_by_vector(embedding.tolist() if hasattr(embedding, "tolist") else embedding, top_k, include_distances=True)
        results = []
        for i, d in zip(ids, distances):
            meta = self._id_to_meta.get(i, {})
            results.append({"id": i, "score": float(d), "text": meta.get("text"), "source_id": meta.get("source_id")})
        return results

File: src/store.js
const fs = require('fs');
const path = require('path');

function cosine(a, b) {
  let s = 0,
    na = 0,
    nb = 0;
  for (let i = 0; i < a.length; i++) {
    s += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  return s / (Math.sqrt(na) * Math.sqrt(nb) || 1);
}

class VectorStore {
  constructor(indexPath) {
    this.indexPath = indexPath;
    this.items = [];
    this._load();
  }

  _load() {
    if (fs.existsSync(this.indexPath)) {
      try {
        const idx = JSON.parse(fs.readFileSync(this.indexPath, 'utf8'));
        this.items = idx.items || [];
      } catch (e) {
        this.items = [];
      }
    }
  }

  _save() {
    fs.writeFileSync(this.indexPath, JSON.stringify({ items: this.items }, null, 2), 'utf8');
  }

  async addChunks(chunks, sourceId, embeddings) {
    const texts = chunks.map((c) => c.text);
    const vecs = await embeddings.embed(texts);
    for (let i = 0; i < chunks.length; i++) {
      this.items.push({
        vector: vecs[i],
        text: chunks[i].text,
        source_id: sourceId,
        page: chunks[i].page,
      });
    }
    this._save();
  }

  async query(q, k, embeddings) {
    const qv = (await embeddings.embed(q))[0];
    const scored = this.items.map((it) => ({ score: cosine(qv, it.vector), meta: { text: it.text, source_id: it.source_id, page: it.page } }));
    scored.sort((a, b) => b.score - a.score);
    return scored.slice(0, k);
  }
}

module.exports = VectorStore;
